Mathlib/Algebra/Lie/InvariantForm.lean

```diff
@@ -0,0 +1,207 @@
+/-
+Copyright (c) 2024 Johan Commelin. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Johan Commelin
+-/
+import Mathlib.Algebra.Lie.Semisimple.Defs
+import Mathlib.LinearAlgebra.BilinearForm.Orthogonal
+
+/-!
+# Lie algebras with non-degenerate invariant bilinear forms are semisimple
+
+In this file we prove that a finite-dimensional Lie algebra over a field is semisimple
+if it does not have non-trivial abelian ideals and it admits a
+non-degenerate reflexive invariant bilinear form.
+Here a form is *invariant* if it is compatible with the Lie bracket: `Φ ⁅x, y⁆ z = Φ x ⁅y, z⁆`.
+
+## Main results
+
+* `LieAlgebra.InvariantForm.orthogonal`: given a Lie submodule `N` of a Lie module `M`,
+  we define its orthogonal complement with respect to a non-degenerate invariant bilinear form `Φ`
+  as the Lie ideal of elements `x` such that `Φ n x = 0` for all `n ∈ N`.
+* `LieAlgebra.InvariantForm.isSemisimple_of_nondegenerate`: the main result of this file;
+  a finite-dimensional Lie algebra over a field is semisimple
+  if it does not have non-trivial abelian ideals and admits
+  a non-degenerate invariant reflexive bilinear form.
+
+## References
+
+We follow the short and excellent paper [dieudonne1953].
+-/
+
+namespace LieAlgebra
+
+namespace InvariantForm
+
+section ring
+
+variable {R L M : Type*}
+variable [CommRing R] [LieRing L] [LieAlgebra R L]
+variable [AddCommGroup M] [Module R M] [LieRingModule L M]
+
+variable (Φ : LinearMap.BilinForm R M) (hΦ_nondeg : Φ.Nondegenerate)
+
+variable (L) in
+/--
+A bilinear form on a Lie module `M` of a Lie algebra `L` is *invariant* if
+for all `x : L` and `y z : M` the condition `Φ ⁅x, y⁆ z = -Φ y ⁅x, z⁆` holds.
+-/
+def _root_.LinearMap.BilinForm.lieInvariant : Prop :=
+  ∀ (x : L) (y z : M), Φ ⁅x, y⁆ z = -Φ y ⁅x, z⁆
+
```

Comment:
I think we should add:
```suggestion

lemma _root_.LinearMap.BilinForm.lieInvariant_iff [LieModule R L M] :
    Φ.lieInvariant (L := L) ↔ ∀ x : L, ⁅x, Φ⁆ = 0 := by
  refine ⟨fun h x ↦ ?_, fun h x y z ↦ ?_⟩
  · ext y z
    rw [LieHom.lie_apply, LinearMap.sub_apply, Module.Dual.lie_apply, LinearMap.zero_apply,
      LinearMap.zero_apply, h, sub_self]
  · replace h := LinearMap.congr_fun₂ (h x) y z
    simp only [LieHom.lie_apply, LinearMap.sub_apply, Module.Dual.lie_apply,
      LinearMap.zero_apply, sub_eq_zero] at h
    simp [← h]

```
(which justifies the name "invariant").

Arguably this should even be the definition (though because of the way I have set up the type instances, we would then pick up a redundant `[LieModule R L M]`).